astropy/stats/funcs.py

```diff
@@ -681,15 +681,29 @@ def median_absolute_deviation(a, axis=None):
 
     """
 
-    a = np.array(a, copy=False)
-    a_median = np.median(a, axis=axis)
+    # Check if the array has a mask and if so use np.ma.median
+    # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
+    # for normal arrays should not be done (summary: np.ma.median always
+    # returns an masked array even if the result should be scalar).
+    try:
+        if np.any(a.mask):
```

Comment:
Ah, I think I forgot one vital step there. 

I had in mind that it could process anything with a ``data``, ``mask`` attribute (even just passing in ``NDData`` would work because that has those two attributes) but then I must have forgotten to set: ``a = np.ma.array(a.data, mask=a.mask, copy=False)`` in the end.

But if I'm overthinking this I could just replace it with a code block such as:
```
    if isinstance(a, np.ma.MaskedArray):
        func = np.ma.median
    else:
        func = np.median
        a = np.asanyarray(a)
```